### PR TITLE
Feature/homescreen

### PR DESCRIPTION
**Pull request by AShozi**
Feature/HomeScreen

 Feature/HomeScreen → Dev 
Home Screen
 
Description Template

 ## Description 
please ignore the funky ui I still need to adjust my constraints.
 
## Screenshots/GIFs 
Screenshot 2024-05-16 at 14 25 15

 
## Checklist 
Please review and check the following before submitting this pull request: 
- [x] The code follows the project's coding standards.

### CODE REVIEW

File: phantomgames/Extensions/Imageview.swift
Comment (AShozi) [inline phantomgames/Extensions/Imageview.swift:1]: indentation looks a bit off

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift
Comment (AShozi) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift:1]: where are the marks

File: phantomgames/Home/HomeScreenModel.swift
Comment (AShozi) [inline phantomgames/Home/HomeScreenModel.swift:1]: this should be removed

File: phantomgames/.DS_Store
Comment (simekanidev) [inline phantomgames/.DS_Store:1]: Please remove this file and make sure it is in your gitIgnore

**Review by AShozi — COMMENTED** [review]
needs minor tweaks

File: phantomgames/.DS_Store
Comment (AShozi) [reply phantomgames/.DS_Store:1]: will do so.

File: phantomgames/GameDetails/GameDetailViewController.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        // Do any additional setup after loading the view.
+    }
+    
+
+    /*
+    // MARK: - Navigation
```
Comment (simekanidev) [inline phantomgames/GameDetails/GameDetailViewController.swift]: Please clean up this file

File: phantomgames/Home/Categories/CategoryViewController.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+        // Do any additional setup after loading the view.
+    }
+    
+
+    /*
+    // MARK: - Navigation
+
+    // In a storyboard-based application, you will often want to do a little preparation before navigation
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```
Comment (simekanidev) [inline phantomgames/Home/Categories/CategoryViewController.swift]: And this one

File: phantomgames/Home/CustomCells/CustomCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+//  Created by Aphiwe Shozi on 2024/05/13.
+//
+
+import UIKit
+
+class CustomCollectionViewCell: UICollectionViewCell {
+    
+    @IBOutlet weak var apiImage: UIImageView!
+    
+    @IBOutlet weak var Gametitle: UILabel!
```
Comment (simekanidev) [inline phantomgames/Home/CustomCells/CustomCollectionViewCell.swift]: Private vars

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+//
+
+import UIKit
+import SDWebImage
+
+class CustomHomeTableViewCell: UITableViewCell {
+
+    
+    @IBOutlet weak var icon: UIImageView!
+
```
Comment (simekanidev) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift:12]: Private

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        playButton.addTarget(self, action: #selector(playButtonTapped), for: .touchUpInside)
+
+    }
+    
+    func populateWith(game: Game) {
+         titleLabel.text = game.title
+        
+                 if let imageURL = URL(string: game.thumbnail) {
```
Comment (simekanidev) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift]: Please check indenting here

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    
+    static func HometableViewNib() -> UINib {
+          UINib(nibName: Constants.TableViewIdentifiers.customHomeCellIdentifier, bundle: nil)
+      }
+      
+    
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+
+        // Configure the view for the selected state
```
Comment (simekanidev) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift]: Extra line breaks

File: phantomgames/Home/HomeScreenModel.swift [hunk 7]
```diff
@@ -0,0 +1,15 @@
+//
+//  HomeScreenModel.swift
+//  phantomgames
+//
+//  Created by Aphiwe Shozi on 2024/05/13.
+//
+
+//import Foundation
+//import UIKit
```
Comment (simekanidev) [inline phantomgames/Home/HomeScreenModel.swift:9]: Please delete

File: phantomgames/Constants/Constants.swift [hunk 8]
```diff
@@ -12,10 +12,11 @@ struct Constants {
     struct SegueIdentifiers {
         
         static let loginSegueIdentifier = "MainSegue"
-        static let pageScreenSegue = "pageScreenSegue"
+        static let GameDetailScreenSegue = "GameDetailScreenSegue"
```
Comment (kieran-woodrow) [inline phantomgames/Constants/Constants.swift:15]: GameDetailScreenSegue
Variables start with small letters.
Please stop ignoring swiftLint warnings

File: phantomgames/GameDetails/GameDetailRepository.swift [hunk 9]
```diff
@@ -0,0 +1,8 @@
+//
+//  GameDetailRepository.swift
```
Comment (kieran-woodrow) [inline phantomgames/GameDetails/GameDetailRepository.swift:2]: can delete this empty file if not in use?

File: phantomgames/GameDetails/GameDetailViewController.swift [hunk 10]
```diff
@@ -0,0 +1,19 @@
+//
+//  GameDetailsViewController.swift
+//  phantomgames
+//
+//  Created by Aphiwe Shozi on 2024/05/16.
+//
+
+import UIKit
+
+class GameDetailViewController: UIViewController {
```
Comment (kieran-woodrow) [inline phantomgames/GameDetails/GameDetailViewController.swift:11]: dele if not being used.

File: phantomgames/GameDetails/GameDetailViewController.swift [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+    }
+
```
Comment (kieran-woodrow) [inline phantomgames/GameDetails/GameDetailViewController.swift]: empty lines?
Please stop ignoring swift lint warnings. I'm going to start decining PR's if i see them not being addressed.

File: phantomgames/GameDetails/GameDetailViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,8 @@
+//
+//  GameDetailViewModel.swift
+//  phantomgames
```
Comment (kieran-woodrow) [inline phantomgames/GameDetails/GameDetailViewModel.swift:3]: not being used?

File: phantomgames/Home/Categories/CategoryRepository.swift [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+}
+class CategoryRepository: CategoryRepositoryType {
+    
+    
+    func fetchAPIImage(completion: @escaping (Result<[CategoryResult], APIError>) -> Void) {
+        
+        let dispatchGroup = DispatchGroup()
+        var results: [CategoryResult] = []
+        
+        dispatchGroup.enter()
```
Comment (kieran-woodrow) [inline phantomgames/Home/Categories/CategoryRepository.swift:25]: this is very strange. Why do you need this?

File: phantomgames/Home/Categories/CategoryViewController.swift [hunk 14]
```diff
@@ -0,0 +1,17 @@
+//
+//  CategoryViewController.swift
+//  phantomgames
+//
+//  Created by Aphiwe Shozi on 2024/05/15.
+//
+
+import UIKit
+
+class CategoryViewController: UIViewController {
```
Comment (kieran-woodrow) [inline phantomgames/Home/Categories/CategoryViewController.swift:10]: empty file again??

File: phantomgames/Home/Categories/CategoryViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,8 @@
+//
+//  CategoryViewModel.swift
+//  phantomgames
```
Comment (kieran-woodrow) [inline phantomgames/Home/Categories/CategoryViewModel.swift:3]: again??

File: phantomgames/Home/CustomCells/CustomCollectionViewCell.swift [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+//
+//  Created by Aphiwe Shozi on 2024/05/13.
+//
+
+import UIKit
+
+class CustomCollectionViewCell: UICollectionViewCell {
+    
+    @IBOutlet weak private var apiImage: UIImageView!
+
```
Comment (kieran-woodrow) [inline phantomgames/Home/CustomCells/CustomCollectionViewCell.swift:19]: no marks and empty space between outlets? why??

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+import SDWebImage
+
+class CustomHomeTableViewCell: UITableViewCell {
+    
+    // MARK: IBOutlets
+    
+    @IBOutlet weak private var icon: UIImageView!
+    
+    @IBOutlet weak private var titleLabel: UILabel!
+
```
Comment (kieran-woodrow) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift]: why are there new lines between outlets??

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 18]
```diff
@@ -0,0 +1,62 @@
+//
+//  CustomHomeTableViewCell.swift
+//  phantomgames
+//
+//  Created by Aphiwe Shozi on 2024/05/14.
+//
+
+import UIKit
+import SDWebImage
```
Comment (kieran-woodrow) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift:9]: Why is this still being used??

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            icon.sd_setImage(with: imageURL, placeholderImage: Constants.ImageConstants.placeholder)
+        } else {
+            icon.image = Constants.ImageConstants.placeholder
+        }
+        if let gameURL = URL(string: game.gameURL) {
+            self.gameURL = gameURL
+        }
+    }
+    
+    @IBAction func playButtonTapped(_ sender: Any) {
```
Comment (kieran-woodrow) [inline phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift]: private...

File: phantomgames/Home/HomeScreenRepository.swift [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    }
+    
+    func fetchHomeResults(completion: @escaping (HomeScreenResult)) {
+        URLSession.shared.request(endpoint: Constants.Endpoints.search, method: .GET, completion: completion)
+    }
+    
+}
+
+
```
Comment (kieran-woodrow) [inline phantomgames/Home/HomeScreenRepository.swift]: a whole bunch of new lines???
Please FIX the swiftlint warnings. I will be declining future PR's. I can't keep commenting this on every PR. Please pss the message to the team as well

File: phantomgames/Home/HomeScreenViewController.swift [hunk 21]
```diff
@@ -5,8 +5,88 @@
...
 
-import Foundation
 import UIKit
 
-class HomeScreenViewController: UIViewController {
+class HomeScreenViewController: UIViewController{
+    
+    // MARK: IBOutlets
+    
+    @IBOutlet weak var HomeCollectionView: UICollectionView!
```
Comment (kieran-woodrow) [inline phantomgames/Home/HomeScreenViewController.swift]: PRIVATE??
I cannot keep commenting on stuff like this anymore

File: phantomgames/Home/HomeScreenViewController.swift [hunk 22]
```diff
@@ -5,8 +5,88 @@
...
+// MARK:  TableView Delegate
+
+extension HomeScreenViewController: UITableViewDelegate,UITableViewDataSource {
+    
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 3
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return viewModel.allGameList.count
```
Comment (kieran-woodrow) [inline phantomgames/Home/HomeScreenViewController.swift]: return statement?? why? Swiftlint says remove the implicit return?

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]

File: phantomgames/GameDetails/GameDetailRepository.swift (hunk 9 above)
Comment (AShozi) [reply phantomgames/GameDetails/GameDetailRepository.swift:2]: Please note that these files are here so that I could create the features segues but the feature development is in the next pull request

File: phantomgames/GameDetails/GameDetailViewController.swift (hunk 10 above)
Comment (AShozi) [reply phantomgames/GameDetails/GameDetailViewController.swift:11]: this goes for all GameDetails page and Category files

File: phantomgames/Home/Categories/CategoryRepository.swift (hunk 13 above)
Comment (AShozi) [reply phantomgames/Home/Categories/CategoryRepository.swift:25]: in the warning I stated that it falls under concurrency so it will be further developed after the workshop

File: phantomgames/Home/Categories/CategoryViewController.swift (hunk 14 above)
Comment (AShozi) [reply phantomgames/Home/Categories/CategoryViewController.swift:10]: I put a warning in the repository file, in future should I put a warning on every file associated with the feature

File: phantomgames/GameDetails/GameDetailRepository.swift (hunk 9 above)
Comment (kieran-woodrow) [reply phantomgames/GameDetails/GameDetailRepository.swift:2]: okay. But please put warnings then. This was all communicated to do so in the PR guidelines and standards

File: phantomgames/Home/CustomCells/CustomCollectionViewCell.swift (hunk 16 above)
Comment (AShozi) [reply phantomgames/Home/CustomCells/CustomCollectionViewCell.swift:19]: my apologies I misunderstood. Thought marks were for organising code for readability didn't know it applied to such short code. Will apply the changes

File: phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift (hunk 18 above)
Comment (AShozi) [reply phantomgames/Home/CustomCells/CustomHomeTableViewCell.swift:9]: I have not been using it in my newer features and I am currently prioritising features that need to be developed not those that have been completed. I will remove it but its not a priority at the moment

File: phantomgames/Home/Categories/CategoryRepository.swift [hunk 23]
```diff
@@ -14,11 +14,9 @@ typealias CategoryResult = (String, Result<[Game], APIError>)
...
     func fetchAPIImage(completion: @escaping (Result<[CategoryResult], APIError>) -> Void)
     func fetchCategory(endpoint: String, completion: @escaping (Result<[Game], APIError>) -> Void)
-
 }
 class CategoryRepository: CategoryRepositoryType {
     
-    
     func fetchAPIImage(completion: @escaping (Result<[CategoryResult], APIError>) -> Void) {
         
         let dispatchGroup = DispatchGroup()
```
Comment (kieran-woodrow) [inline phantomgames/Home/Categories/CategoryRepository.swift:22]: why is this needed?

File: phantomgames/Home/HomeScreenViewController.swift [hunk 24]
```diff
@@ -11,8 +11,8 @@ class HomeScreenViewController: UIViewController{
     
     // MARK: IBOutlets
     
-    @IBOutlet weak var HomeCollectionView: UICollectionView!
-    @IBOutlet weak var tableView: UITableView!
+    @IBOutlet weak private var HomeCollectionView: UICollectionView!
```
Comment (kieran-woodrow) [inline phantomgames/Home/HomeScreenViewController.swift]: why is this variable starting with a capital letter?
Please read through the recommended standards that we have spend 4 lessons on and made documents for.

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]

File: phantomgames/Home/Categories/CategoryRepository.swift (hunk 23 above)
Comment (AShozi) [reply phantomgames/Home/Categories/CategoryRepository.swift:22]: for my filter by category feature. Each category has its own endpoint and there's at least 6 categories so the calls are organised using a dispatch group

Comment (AShozi) [reply phantomgames/Home/Categories/CategoryRepository.swift:22]: as stated in my warning it part of concurrency. so I left the repo in a state where it has no errors and im hoping the concurrency workshop will give more clarity on developing the feature

**Review by simekanidev — APPROVED** [review]

**Review by kieran-woodrow — APPROVED** [review]
